feat: instrument the package

y1yang0 opened the pull request:
This patch introduces the real implementation of the instrumentation

It looks something like this
```bash

$ make demo
Packaging instrumentation code into binary...
Building instrumentation tool...
otel version v0.0.0-20250905015142-26f1d64fe8e5+26f1d64fe8e5ca7434b6fc32c95d126b84f2a5fa (2025-09-05T01:51:42Z)
Building demo with instrumentation...
go: found github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/instrumentation/helloworld in github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/instrumentation/helloworld v0.0.0-00010101000000-000000000000
Running demo...
=setupOpenTelemetry=
[MyHook] start to instrument hello world!
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:demo"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.37.0"}}],"ScopeMetrics":[]}
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:demo"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.37.0"}}],"ScopeMetrics":[]}
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:demo"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.37.0"}}],"ScopeMetrics":[]}
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:demo"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.37.0"}}],"ScopeMetrics":[]}
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:demo"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.37.0"}}],"ScopeMetrics":[]}
{"Name":"hello-world","SpanContext":{"TraceID":"e7eda03eb4bd4d1543b0a55274c6eef2","SpanID":"b4bc176cf9ac82fc","TraceFlags":"01","TraceState":"","Remote":false},"Parent":{"TraceID":"00000000000000000000000000000000","SpanID":"0000000000000000","TraceFlags":"00","TraceState":"","Remote":false},"SpanKind":1,"StartTime":"2025-09-05T09:52:25.273622243+08:00","EndTime":"2025-09-05T09:52:30.274611445+08:00","Attributes":[{"Key":"url.scheme","Value":{"Type":"STRING","Value":"http"}},{"Key":"url.path","Value":{"Type":"STRING","Value":"/a"}},{"Key":"url.query","Value":{"Type":"STRING","Value":"a=5"}}],"Events":null,"Links":null,"Status":{"Code":"Unset","Description":""},"DroppedAttributes":0,"DroppedEvents":0,"DroppedLinks":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:demo"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.37.0"}}],"InstrumentationScope":{"Name":"hello-world","Version":"0.0.1","SchemaURL":"","Attributes":null},"InstrumentationLibrary":{"Name":"hello-world","Version":"0.0.1","SchemaURL":"","Attributes":null}}
[MyHook] hello world is instrumented!
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:demo"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.37.0"}}],"ScopeMetrics":[{"Scope":{"Name":"hello-world","Version":"","SchemaURL":"","Attributes":null},"Metrics":[{"Name":"http.client.request.duration","Description":"Duration of HTTP client requests.","Unit":"ms","Data":{"DataPoints":[{"Attributes":[{"Key":"url.scheme","Value":{"Type":"STRING","Value":"http"}}],"StartTime":"2025-09-05T09:52:25.273612571+08:00","Time":"2025-09-05T09:52:31.274312009+08:00","Count":1,"Bounds":[0,5,10,25,50,75,100,250,500,750,1000,2500,5000,7500,10000],"BucketCounts":[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"Min":-9223372036854776000,"Max":-9223372036854776000,"Sum":-9223372036854776000,"Exemplars":[{"FilteredAttributes":null,"Time":"2025-09-05T09:52:30.274844931+08:00","Value":-9223372036854776000,"SpanID":"tLwXbPmsgvw=","TraceID":"5+2gPrS9TRVDsKVSdMbu8g=="}]}],"Temporality":"CumulativeTemporality"}}]}]}
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:demo"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.37.0"}}],"ScopeMetrics":[{"Scope":{"Name":"hello-world","Version":"","SchemaURL":"","Attributes":null},"Metrics":[{"Name":"http.client.request.duration","Description":"Duration of HTTP client requests.","Unit":"ms","Data":{"DataPoints":[{"Attributes":[{"Key":"url.scheme","Value":{"Type":"STRING","Value":"http"}}],"StartTime":"2025-09-05T09:52:25.273612571+08:00","Time":"2025-09-05T09:52:32.274525757+08:00","Count":1,"Bounds":[0,5,10,25,50,75,100,250,500,750,1000,2500,5000,7500,10000],"BucketCounts":[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"Min":-9223372036854776000,"Max":-9223372036854776000,"Sum":-9223372036854776000,"Exemplars":[{"FilteredAttributes":null,"Time":"2025-09-05T09:52:30.274844931+08:00","Value":-9223372036854776000,"SpanID":"tLwXbPmsgvw=","TraceID":"5+2gPrS9TRVDsKVSdMbu8g=="}]}],"Temporality":"CumulativeTemporality"}}]}]}
[MyHook] after hook executed!
```

y1yang0 commented on tool/cmd/cmd-go.go:
```diff
@@ -34,12 +42,12 @@ var commandGo = cli.Command{
 		err = setup.Setup(ctx)
 		if err != nil {
-			return cli.Exit(err, exitCodeFailure)
+			return ex.Errorf(err, "failed to build with toolexec with exitcode %d", exitCodeFailure)
```
Side note: cli.exit was changed to ex.error to preserve the error chain. Otherwise, ex.fatal wouldn't be able to print the correct error stack.

y1yang0 replied:
See below

pdelewski commented on tool/internal/instrument/apply_func.go:
```diff
@@ -0,0 +1,321 @@
+				stmt.Decorations().Start.Append(tag)
+			}
+		} else {
+			pos = ip.parser.FindPosition(funcDecl.Body)
```
This line seems redundant as pos is computed before if statement 
```suggestion
```

pdelewski commented on tool/internal/instrument/apply_func.go:
```diff
@@ -0,0 +1,321 @@
+	ip.insertToFunc(funcDecl, tjump)
+
+	// Generate corresponding trampoline code
+	err := ip.generateTrampoline(t)
```
Can we simplify that to just 
```suggestion
	return ip.generateTrampoline(t)
```

pdelewski replied:
and remove 4 below lines

pdelewski commented on tool/internal/setup/add.go:
```diff
@@ -13,19 +15,97 @@ const (
+		// First variable declaration
+		// //go:linkname _getstatck%d %s.OtelGetStackImpl
+		// var _getstatck%d = _otel_debug.Stack
+		value := ast.SelectorExpr(ast.Ident("_otel_debug"), "Stack")
```
Similar as above

pdelewski commented on tool/internal/ast/shared.go:
```diff
@@ -8,16 +8,52 @@ import (
+	return decls[0], nil
+}
+
+func FindFuncDeclWithRecv(root *dst.File, funcName string) ([]*dst.FuncDecl, error) {
```
The current name is misleading IMO: it returns both free functions and methods. Maybe rename it to:
```suggestion
func FindFuncDecl(root *dst.File, funcName string) ([]*dst.FuncDecl, error) {
```

pdelewski commented on tool/internal/ast/shared.go:
```diff
@@ -8,16 +8,52 @@ import (
+	return found, nil
+}
+
+func FindFuncDecl(root *dst.File, funcName string) (*dst.FuncDecl, error) {
```
Rename it to (look my comment below):
```suggestion
func FindFuncDeclWithoutRecv(root *dst.File, funcName string) (*dst.FuncDecl, error) {
```

kakkoyun reviewed (commented):
I did an initial pass and I hope to have another pass soon.

pdelewski commented:
I’d like to merge this soon, as I have follow-up changes depending on it and more planned.

y1yang0 commented on tool/data/export.go:
```diff
@@ -9,7 +9,7 @@ import (
 	"github.com/open-telemetry/opentelemetry-go-compile-instrumentation/tool/ex"
 )
 
-//go:embed *.yaml
```
Because we need both *.yaml and *.gzip

kakkoyun commented:
@RomainMuller was on PTO and just returned. Let's give him a chance to review it and then we can proceed.

RomainMuller commented on tool/internal/setup/extract.go:
```diff
@@ -0,0 +1,143 @@
+}
+
+func extractGZip(data []byte, targetDir string) error {
+	err0 := os.MkdirAll(targetDir, 0o755)
+	if err0 != nil {
+		return ex.Error(err0)
+	}
+
+	gzReader, err0 := gzip.NewReader(strings.NewReader(string(data)))
+	if err0 != nil {
+		return ex.Error(err0)
+	}
+	defer func() {
+		err0 = gzReader.Close()
+		if err0 != nil {
+			ex.Fatal(err0)
+		}
+	}()
```
Why are these not named `err` but `err0`? That looks unidiomatic?

y1yang0 replied:
I renamed the first error to err0 because the name err is used for multiple subsequent checks in this function. This was primarily for brevity, so I could keep using the simple err variable. My apologies for the slightly awkward err0 name!

y1yang0 replied:
Sounds good, I have no preference here. Feel free to refactor it as you see fit.

RomainMuller commented on tool/internal/setup/extract.go:
```diff
@@ -0,0 +1,143 @@
+		}
+
+		// Skip AppleDouble files (._filename) and other hidden files
+		if strings.HasPrefix(filepath.Base(header.Name), "._") ||
+			strings.HasPrefix(filepath.Base(header.Name), ".") {
```
Strings that start with `._` all start with `.` so it is pointless to check both.

```suggestion
		if strings.HasPrefix(filepath.Base(header.Name), ".") {
```

RomainMuller commented on tool/util/go.go:
```diff
@@ -85,3 +85,7 @@ func SplitCompileCmds(input string) []string {
 }
+
+func IsGoFile(path string) bool {
+	return strings.HasSuffix(path, ".go")
```
That is susceptible to false negatives on case-insensitive file systems.

RomainMuller commented on .golangci.yml:
```diff
@@ -197,7 +197,7 @@ linters:
       # Such cases aren't reported by default.
       # Default: false
-      check-type-assertions: true
+      check-type-assertions: false
```
Why so? Not doing it and making a mistake results in a panic... this is kind of dangerous?

RomainMuller reviewed (changes requested):
I was going to mention that we should probably to to work without `dave/dst` but I realize it makes it difficult to manage the `//go:linkname` directives that we need to output. That being said, I remain convinced we'd be better off without `dave/dst` in the long run because that dependency is not super well maintained anymore (Dave admitted to not having a lot of time to allocate to it since a long while). It also has a number of fairly bad bugs when dealing with import lists that cause hard failures when used against arbitrary code.

---

Second, this PR introduces large chunks of functionality without introducing corresponding unit tests. We need to improve our test coverage game or we will end up with future maintainability hazards as the product evolves.

---

Finally, the codebase will be more approachable to new maintainers/contributors if - at minimum - all exported functions had adequate documentation strings. This would at least make it easier to understand some of the intents without having to reason over the entire code.

kakkoyun commented:
> I was going to mention that we should probably to to work without `dave/dst` but I realize it makes it difficult to manage the `//go:linkname` directives that we need to output. That being said, I remain convinced we'd be better off without `dave/dst` in the long run because that dependency is not super well maintained anymore (Dave admitted to not having a lot of time to allocate to it since a long while). It also has a number of fairly bad bugs when dealing with import lists that cause hard failures when used against arbitrary code.
> 
> Second, this PR introduces large chunks of functionality without introducing corresponding unit tests. We need to improve our test coverage game or we will end up with future maintainability hazards as the product evolves.
> 
> Finally, the codebase will be more approachable to new maintainers/contributors if - at minimum - all exported functions had adequate documentation strings. This would at least make it easier to understand some of the intents without having to reason over the entire code.

Thanks, Romain, for raising the bar here. Starting with a high standard of quality pays dividends down the line. If we need to go deeper on any of the points, we can always set up an ad-hoc chat since our weekly sync was already this morning.

pdelewski commented:
@y1yang0 I extracted `compileCommand` from `toolexec` and added `instrument_test.go`. I pushed these changes to your branch so we don’t lose them and to keep it in sync with the current implementation. Feel free to tweak or revert if you prefer.

pdelewski commented on tool/internal/instrument/apply_func.go:
```diff
@@ -0,0 +1,339 @@
+	OtelGlobalsFile = "otel.globals.go"
+)
+
+func makeName(r *rule.InstFuncRule,
+	funcDecl *dst.FuncDecl, isBefore bool,
+) string {
```
That might be good idea

kakkoyun commented:
> I was going to mention that we should probably to to work without `dave/dst` but I realize it makes it difficult to manage the `//go:linkname` directives that we need to output. That being said, I remain convinced we'd be better off without `dave/dst` in the long run because that dependency is not super well maintained anymore (Dave admitted to not having a lot of time to allocate to it since a long while). It also has a number of fairly bad bugs when dealing with import lists that cause hard failures when used against arbitrary code.

Also, not using dave/dst is news to me. I agree it’s a good direction in the long run, however I think it’s more of a stretch goal at this stage. For now, We can keep the focus on getting the basics right first. There’s also an upstream Go issue that might help resolve this in the future: https://github.com/golang/go/issues/20744.

Besides that we should be any architectural decision that would tight our hands in the future.

y1yang0 commented on tool/internal/ast/parser.go:
```diff
@@ -48,8 +48,23 @@ func (ap *AstParser) ParseFile(filePath string, mode parser.Mode) (*dst.File, er
+// ParseSource parses the AST from complete source code.
+func (ap *AstParser) ParseSource(source string) (*dst.File, error) {
+	util.Assert(source != "", "empty source")
+	ap.dec = decorator.NewDecorator(ap.fset)
```
Good catch

y1yang0 commented on tool/internal/ast/primitives.go:
```diff
@@ -0,0 +1,221 @@
+	}
+}
+
+func Exprs(exprs ...dst.Expr) []dst.Expr {
+	return exprs
+}
+
+func Stmts(stmts ...dst.Stmt) []dst.Stmt {
+	return stmts
+}
```
For consistency see above

y1yang0 commented on tool/internal/instrument/toolexec.go:
```diff
@@ -6,51 +6,112 @@ package instrument
+		}
 	}
+	return args
+}
```
We will insert function with no body for further golinkname, I'll add a comment

```
func MyHookBefore(hookContext HookContext)
func MyHookAfter(hookContext HookContext)

```

y1yang0 commented on tool/internal/instrument/toolexec.go:
```diff
@@ -6,51 +6,112 @@ package instrument
-			return err
+		// Read compilation output directory
+		target := util.FindFlagValue(args, "-o")
+		util.Assert(target != "", "why not otherwise")
```
This ensures that we can always parse the -o flag from the compile command. The goal is to prevent any unexpected behavior or nasty surprises from surfacing later in the code.

y1yang0 commented on tool/internal/instrument/toolexec.go:
```diff
@@ -6,51 +6,112 @@ package instrument
-			// return nil
+
+			// Strip -complete flag as we may insert some hook points that are
+			// not ready yet, i.e. they don't have function body
```
The -complete flag is required when there is no function body. Removing it will cause an error
```
Building demo with instrumentation...
go: found github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/instrumentation/helloworld in github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/instrumentation/helloworld v0.0.0-00010101000000-000000000000
WORK=/tmp/go-build1863349319
# github.com/open-telemetry/opentelemetry-go-compile-instrumentation/demo
<generated>:112: missing function body
<generated>:113: missing function body
Error:
[0] failed to run command /usr/local/go/pkg/tool/linux_amd64/compile: exit status 2
[1] failed to run toolexec with exitcode -1

Stack:
[0]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/util/sys.go:30 /tool/util.RunCmdWithEnv
[1]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/util/sys.go:36 /tool/util.RunCmd
[2]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/internal/instrument/toolexec.go:116 /tool/internal/instrument.Toolexec
[3]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/cmd/cmd-toolexec.go:22 main.init.func3
[4]/home/y1yang0/go/pkg/mod/github.com/urfave/cli/v3@v3.4.1/command_run.go:344 github.com/urfave/cli/v3.(*Command).run
[5]/home/y1yang0/go/pkg/mod/github.com/urfave/cli/v3@v3.4.1/command_run.go:278 github.com/urfave/cli/v3.(*Command).run
[6]/home/y1yang0/go/pkg/mod/github.com/urfave/cli/v3@v3.4.1/command_run.go:94 github.com/urfave/cli/v3.(*Command).Run
[7]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/cmd/main.go:52 main.main
[8]/usr/local/go/src/runtime/proc.go:283 runtime.main
Error:
[0] failed to run command go: exit status 1
[1] failed to build with toolexec with exitcode -1

Stack:
[0]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/util/sys.go:30 /tool/util.RunCmdWithEnv
[1]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/internal/setup/setup.go:117 /tool/internal/setup.BuildWithToolexec
[2]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/cmd/cmd-go.go:46 main.init.func1
[3]/home/y1yang0/go/pkg/mod/github.com/urfave/cli/v3@v3.4.1/command_run.go:344 github.com/urfave/cli/v3.(*Command).run
[4]/home/y1yang0/go/pkg/mod/github.com/urfave/cli/v3@v3.4.1/command_run.go:278 github.com/urfave/cli/v3.(*Command).run
[5]/home/y1yang0/go/pkg/mod/github.com/urfave/cli/v3@v3.4.1/command_run.go:94 github.com/urfave/cli/v3.(*Command).Run
[6]/home/y1yang0/opentelemetry-go-compile-instrumentation/tool/cmd/main.go:52 main.main
[7]/usr/local/go/src/runtime/proc.go:283 runtime.mainmake: *** [Makefile:44: demo] Error 1
```

y1yang0 commented on tool/internal/setup/add.go:
```diff
@@ -13,19 +15,99 @@ const (
-	importDecls := make([]dst.Decl, 0, len(matched))
+//nolint:gochecknoglobals // This is a constant
+var requiredImports = [][]string{
+	{"runtime/debug", "_otel_debug"}, // The getstack function depends on runtime/debug
+	{"log", "_otel_log"},             // The printstack function depends on log
```
Yes, always

y1yang0 commented on tool/internal/ast/primitives.go:
For consistency see above

y1yang0 commented on tool/internal/instrument/apply_func.go:
```diff
@@ -0,0 +1,339 @@
+		prefix = TrampolineBeforeName
+	}
+	return fmt.Sprintf("%s_%s%s",
+		prefix, funcDecl.Name.Name, util.CRC32(r.String()))
```
A name conflict here only occurs in one specific scenario: when the same function is instrumented by multiple rules. In that case, we need to add a suffix to the before and after functions for each rule to prevent the generated trampoline function names from colliding.

However, I believe a simple and short CRC32 hash is sufficient to prevent this, for two main reasons:

- It's a rare edge case: The situation where a single function is targeted by multiple instrumentation rules is expected to be very uncommon in practice.
- Hashes will differ: The CRC32 hash is calculated based on the rule's content, so different rules will naturally produce different hash values.

Given these points, the probability of a collision is extremely low, and a short CRC32 provides an adequate and efficient solution.

RomainMuller replied:
Maybe we should then make it opaque, so we can chang it later on (if needed) without breaking untold assumptions... I.e, rename it to `util.Checksum` or `util.Hash` or something?

y1yang0 commented on Makefile:
```diff
@@ -15,9 +17,20 @@ BUILD_TIME := $(shell date -u '+%Y-%m-%d')
 
+# Package the instrumentation code into binary
+.PHONY: package
+package:
+	@echo "Packaging instrumentation code into binary..."
+	@rm -rf $(INST_PKG_TMP)
+	@cp -a pkg $(INST_PKG_TMP)
+	@cd $(INST_PKG_TMP) && go mod tidy
+	@tar -czf $(INST_PKG_GZIP) --exclude='*.log' $(INST_PKG_TMP)
+	@mv $(INST_PKG_GZIP) tool/data/
+	@rm -rf $(INST_PKG_TMP)
```
There are several viable solutions for this, but let's defer addressing it for now, maybe go-lize the Makefile

y1yang0 commented:
Thanks, everyone, for the incredibly helpful feedback and thorough review.

---

> Second, this PR introduces large chunks of functionality without introducing corresponding unit tests. We need to improve our test coverage game or we will end up with future maintainability hazards as the product evolves.

@RomainMuller  Yes, this is a high-priority item on my to-do list, and I'll be adding more tests in follow-up PRs.

> Finally, the codebase will be more approachable to new maintainers/contributors if - at minimum - all exported functions had adequate documentation strings. This would at least make it easier to understand some of the intents without having to reason over the entire code.

@RomainMuller I'll add the documentation to this PR to make the codebase more approachable for new contributors.

---

> For some reason I was thinking that we discussed that before and there is an internal agreement to use dave/dst. Personally, I would like to use standard library, however I know that this library has some advantages.

@pdelewski  Yes, we have discussed this topic before. Let's agree to defer the dave/dst discussion for now and revisit it later.

> I agree with you and we discussed this topic on last SIG meeting. My idea was to merge this sooner than later and then focus on testability.

@pdelewski  Yes, agreed. Adding more tests is a high priority for me as well.

> I extracted compileCommand from toolexec and added instrument_test.go. I pushed these changes to your branch so we don’t lose them and to keep it in sync with the current implementation. I originally planned to do that in separate PR. Feel free to tweak or revert if you prefer.

@pdelewski  Thank you for adding this code! I think it's valuable and highly relevant to the scope of this PR. Let's definitely keep it.

---

> Also, not using dave/dst is news to me. I agree it’s a good direction in the long run, however I think it’s more of a stretch goal at this stage. For now, We can keep the focus on getting the basics right first.

@kakkoyun Yes. Let's focus on getting the basics right in this PR and park the dave/dst discussion for a future effort.

y1yang0 commented on tool/data/export.go:
```diff
@@ -9,7 +9,7 @@ import (
 )
 
-//go:embed *.yaml
+//go:embed *
```
The otel-pkg.gz is generated before compile-time, and fixing this is non-trivial. I'll address it in a follow-up PR. Tracked in #29

ralf0131 commented on pkg/inst/context.go:
```diff
@@ -3,7 +3,12 @@
 package inst
 
-type Context interface {
+// !!!Any modification MUST be synced with tool/internal/instrument/template_api.go
```
```suggestion
// !!! Any modification MUST be synced with tool/internal/instrument/template_api.go
```

ralf0131 commented on tool/cmd/cmd-go.go:
```diff
@@ -34,12 +40,12 @@ var commandGo = cli.Command{
 		err = setup.Setup(ctx)
 		if err != nil {
-			return cli.Exit(err, exitCodeFailure)
+			return ex.Errorf(err, "failed to build with toolexec with exitcode %d", exitCodeFailure)
```
```suggestion
			return ex.Errorf(err, "failed to build with toolexec with exit code %d", exitCodeFailure)
```

ralf0131 commented on tool/cmd/cmd-go.go:
```diff
@@ -34,12 +40,12 @@ var commandGo = cli.Command{
 		err = setup.BuildWithToolexec(ctx, cmd.Args().Slice())
 		if err != nil {
-			return cli.Exit(err, exitCodeFailure)
+			return ex.Errorf(err, "failed to build with toolexec with exitcode %d", exitCodeFailure)
```
```suggestion
			return ex.Errorf(err, "failed to build with toolexec with exit code %d", exitCodeFailure)
```

ralf0131 commented on tool/cmd/cmd-setup.go:
```diff
@@ -18,7 +19,7 @@ var commandSetup = cli.Command{
 		err := setup.Setup(ctx)
 		if err != nil {
-			return cli.Exit(err, exitCodeFailure)
+			return ex.Errorf(err, "failed to setup with exitcode %d", exitCodeFailure)
```
```suggestion
			return ex.Errorf(err, "failed to setup with exit code %d", exitCodeFailure)
```

kakkoyun commented on pkg/inst/context.go:
```diff
@@ -3,7 +3,12 @@
+	// Get the skip call flag, can be used to skip the original function call
+	IsSkipCall() bool
 	// Set the data field, can be used to pass information between Before and After hooks
 	SetData(interface{})
```
nit: We should migrate using `any` instead of `interface{}`.

y1yang0 replied:
Let's stick to using interface{} in the generated code for better backward compatibility. This ensures the code will compile correctly in environments where any is not yet defined. In normal tool code, `any` is definitively better and preferable

kakkoyun replied:
If we decide to keep the integrations within the same repo, believe me supporting old versions will be a lot problem. But this is another discussion.

There are linters and a modern go initiative as part of `go vet` will update these sort of patterns. So I don't want to make this a bikeshedding discussion :)

kakkoyun commented on tool/data/helloworld.yaml:
```diff
@@ -1,13 +1,12 @@
-  advice:
-    - before: MyHookBefore
-    - after: MyHookAfter
+  # What hook code we want to inject at target func entry or exit
+  before: MyHookBefore
+  after: MyHookAfter
```
Even though, you can do a lot of things with "before" and "after", it is just now enough. Our experience on writing integrations (https://github.com/search?q=repo%3ADataDog%2Fdd-trace-go%20orchestrion.yml&type=code), showed us you need a lit more flexibility. You can check the abstractions we had to invent previously https://datadoghq.dev/orchestrion/contributing/aspects/join-points

So I don't want us to tie our hands here. In the OpenTelemetry community, I'm sure we will have many more integrations and having the flexibility to expand on this space is quite valuable.

y1yang0 replied:
Got it, thanks for explaining. I see where we differ now. What you call "advice" is basically what I call different "rules" in my design.

My InstRule interface is designed for exactly this, to support multiple rule types. We already have InstFuncRule, and I'm planning to add InstStructRule (for adding struct fields) and InstFileRule (as the most flexible option).

So, while we might disagree on the naming, it seems we're on the same page about the overall direction.

kakkoyun replied:
Sorry, the request was specifically about the "advice"s, so this is more relevant https://datadoghq.dev/orchestrion/contributing/aspects/advice/

kakkoyun reviewed (commented):
More comments.

pdelewski commented:
> I don't want to nitpick each and every thing, not to block the PR, however, I just wanted to note, we will have to change too many things afterwards, especially when it comes to style.
> 
> Some of the methods names are not idiomatic to Go, some of the file names are not very idiomatic to go. We are introducing too many indirections (helper functions, util packages etc.), with good intentions. to make the code more readable, however (again) this is not idiomatic Go. Some of the functions makes me feel like we try to invent a domain-specific language around AST.
> 
> We should also have to be conscience about allocations.

It’s true that some filenames are non-idiomatic and inconsistent (e.g., cmd-go.go vs. func_rule.go). We should use underscores (_) instead of hyphens (-). Some files could be combined to reduce the number of files—for example, instrument.go contains only one function. That said, we should all agree on what needs to be fixed in this PR and what can be deferred as follow-up tasks. I’m okay with merging this as is and making improvements afterward.

ralf0131 commented:
I am also okay with merging it first and improve it later, I think right now it is important for us to make progress rather than making it perfect.

y1yang0 commented:
I've made the following changes:

- Replaced hyphens with underscores in all relevant filenames for consistency.
- Consolidated load.go and match.go into a single match.go file.

The structure of other files remains unchanged to allow for potential future feature additions.
I plan to merge this patch on Friday, Sep 26th if there is no further feedback. Thank you for the review!

RomainMuller commented on tool/internal/ast/primitives.go:
```diff
@@ -0,0 +1,221 @@
+	IdentIgnore = "_"
+)
+
+func Ident(name string) *dst.Ident {
+	return &dst.Ident{
+		Name: name,
+	}
+}
```
It could then be (but this is not a big deal)

```suggestion
func Ident(name string) *dst.Ident {
	return dst.NewIdent(name)
}
```

RomainMuller commented on tool/internal/instrument/template_api.go:
My concern here is really that there are both templates and regular go code in this package, and it's not entirely obvious from looking at the file tree... You might be implying that the `template_` prefix here is how you know what is a template versus code that is built as part of the tool? Then I guess maybe that's fine... although from an aesthetics standpoint, I tend to prefer differentiating based on the file extension rather than prefix :)

RomainMuller commented on tool/internal/instrument/template_impl.go:
```diff
@@ -0,0 +1,119 @@
+package instrument
+
+//line <generated>:1
+type HookContextImpl struct {
+	Params      []interface{}
+	ReturnVals  []interface{}
+	SkipCall    bool
+	Data        interface{}
+	FuncName    string
+	PackageName string
+}
```
That is supporting `go`s before 1.18... Which then also means strictly no use of generics anywhere in the injected code (for `go` before 1.18 will not be able to compile those), etc... And also this is `go` runtimes that long-since out of support...

I don't care all that much, but I personally thing explicitly trying to support these is making a disservice to the Go community by encouraging continued use of dangerously old releases.

y1yang0 commented:
@RomainMuller Thanks for the final review, Romain. I was out of office for past few days. I'll address your comments in follow-up PRs.
